Ensure $oldSessionData is an array before merge

exptom opened the pull request:
If $oldSessionData is a Traversable type of storage rather than a plain array (e.g. `Zend\Session\Storage\SessionStorage` then ArrayUtils::merge would throw an error, Traversable types will be converted to arrays before merge.

This fixes issue #10

exptom committed:
Ensure $oldSessionData is an array before merge

If $oldSessionData is a Traversable type of storage rather than a plain array (e.g. `Zend\Session\Storage\SessionStorage` then ArrayUtils::merge would throw an error, Traversable types will be converted to arrays before merge.

samsonasik commented on src/SessionManager.php:
```diff
@@ -133,6 +133,10 @@ public function start($preserveStorage = false)
         if ($oldSessionData instanceof \Traversable
             || (! empty($oldSessionData) && is_array($oldSessionData))
         ) {
+            if(!is_array($oldSessionData)) {
```
space after if

samsonasik commented on src/SessionManager.php:
```diff
@@ -133,6 +133,10 @@ public function start($preserveStorage = false)
+            if(!is_array($oldSessionData)) {
+                $oldSessionData = $oldSessionData->toArray();
+            }
+            
```
remove whitespace

samsonasik reviewed (changes requested):
cs

samsonasik commented:
also, it seems require tests

weierophinney reviewed (changes requested):
As noted by @samsonasik:

- We need unit tests.
- Please use the provided phpcs configuration to fix CS issues (`composer cs-fix`).

Ocramius commented:
@exptom please don't merge upstream branches into yours: rebase instead.

exptom commented:
@Ocramius Apologies! How can I fix this now?

samsonasik commented:
@exptom I created a blog post about rebase feature branch: https://samsonasik.wordpress.com/2015/09/16/practical-git-4-rebasing-conflicted-task-branch-against-primary-branch/

exptom commented:
@samsonasik Thank you, is there a way to easily revert back the pushed merged so that I can rebase instead?

samsonasik commented:
First, do reset --hard latest commit ( merge commit ): 

```
git reset --hard HEAD~1
```

then, switch to master branch and pull upstream master:

```
git checkout master && git pull https://github.com/zendframework/zend-session.git master
```

then, switch to your branch again and rebase:

```
git checkout patch-1 && git rebase master
```

You will get conflict, then you need to fix the conflict indicating by "<<<<<<< HEAD", don't commit anything at this phase.

After all conflicted files resolved, do:

```
git add . && git rebase --continue && git push --force origin patch-1
```

exptom committed:
Codestyle fixes and unit test

exptom commented:
@samsonasik I think I have fixed it
@weierophinney Thanks for the reminder, I had forgotten to come back to this ticket.
I have applied the CS fixes and added a unit test.

I have had to make a change though as the setUp() method was creating a new instance of SessionManager and that was causing problems testing the storage. When the SessionManager instance is created it registers a shutdown function but because I need to create a new SessionManager to test the particular storage object it was still causing the shutdown function registered in the setUp() to run and causing the test to fail. I have moved the SessionManager instantiation into the individual tests.

roelvanduijnhoven commented:
Also encounter this while upgradinging from ZF2 to ZF3. Not sure how to proceed. For the time being I'll use the following *fix*. But obviously not valid for production:

> 'type' => \Zend\Session\Storage\ArrayStorage::class,

zf2timo commented:
I encounter this issue with the `\Zend\Session\Storage\SessionArrayStorage::class` today.
The suggested workaround from @roelvanduijnhoven did not worked. I still get the error:
> Argument 1 passed to Zend\Stdlib\ArrayUtils::merge() must be of the type array, object given, called in /var/www/html/vendor/zendframework/zend-session/src/SessionManager.php on line 136

I think this PR should be merged in the near feature, because it is already opened for over a year and it occurs when the example configuration from the documentation is used.

Xerkus commented on src/SessionManager.php:
```diff
@@ -133,6 +133,10 @@ public function start($preserveStorage = false)
             || (! empty($oldSessionData) && is_array($oldSessionData))
         ) {
+            if(! is_array($oldSessionData)) {
+                $oldSessionData = $oldSessionData->toArray();
```
This assumes $oldSessionsData from $_SESSION superglobal is an object that implements `toArray()` method.
Can it be something else?

weierophinney commented:
Thanks, @exptom
